rfc/0010-cbft.md

```diff
@@ -0,0 +1,1355 @@
+
+Range Queries:
+
+ - **Date Range Query:** The date range query finds documents containing a date value in the specified field within the specified range.
+ - **Numeric Range Query:** The numeric range query finds documents containing a numeric value in the specified field within the specified range.
+
+Compound Queries:
+
+ - **Conjunction Query:** Result documents must satisfy all of the child queries.
+ - **Disjunction Query:** Result documents must satisfy a configurable min number of child queries.
+ - **Boolean Query:** The boolean query is a useful combination of conjunction and disjunction queries.
+
+Some queries are more limited and usually used for debugging purposes:
+
+ - **TermQuery:** A query that looks for **exact** matches of the term in the index (no analyzer, no stemming). Useful to check what the actual content of the index is. It can also apply fuzziness on the term. Usual better alternative is `MatchQuery`.
+ - **Prefix Query:** The prefix query finds documents containing terms that start with the provided prefix. Usual better alternative is `MatchQuery`.
+ - **PhraseQuery:** A query that looks for **exact** match of several terms (in the exact order) in the index. Usual better alternative is `MatchPhraseQuery`.
+ - **MatchAllQuery:** A query that matches all indexed documents.
+ - **MatchNoneQuery:** A query that matches nothing.
+
+## 2. High Level API Overview
+This section describes how the overall API looks like.
+
+The search service is an HTTP endpoint that accepts `POST` requests. The url of the endpoint determine the FTS index to be queried, while the body of the request contains a `JSON` representation of the FTS query to perform:
+
+> **POST** /api/index/`{indexName}`/query
+
+For both bucket and cluster level, the approach is the same. Right now only Bucket level is defined in this RFC.
+
+```java
+bucket.query(String indexName, SearchQuery query[, SearchParams params]);
+```
+
+All query types described below must implement the `SearchQuery` marker interface or base class (including compound queries since they can be nested too).
+
+`SearchQuery` are composites (you can nest queries inside some specific query types), and the search parameters (like limit, explain) are only expected at top level of the payload, hence the separation of `SearchQuery` and the optional `SearchParams`.
+
+Here is an example of a String query with some search params, on index `beer-index`:
+
+```java
+bucket.query("beer-index", new StringQuery("description:awesome").boost(2), new SearchParams().limit(5))
+```
+
+If idiomatic by the language, it makes sense to provide static factory methods (or similar helpers) to guide the user in the right direction since there are many different query options available. So for example the `SearchQuery` can contain factory methods in a form like this:
+
+```java
+bucket.query("beer-index", SearchQuery.string("+water +abv:>10"), new SearchParams().limit(5))
```

Comment:
it makes sense if you see the `SearchQuery` as the equivalent of the N1QL statement. And when you consider that composing queries (eg. in a `DisjunctionQuery`) has the strong requirement that these subqueries should **not** have any searchParams. So you create the JSON request by taking the `indexName` (for the URL) and both `SearchQuery` and `SearchParams` for the body.